src/compiler/checker.ts

```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         for (let i = 0; i < paramCount; i++) {
             const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-            const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            if (i === restIndex && targetType && sourceType && isTupleType(sourceType)) {
+                targetType = mapType(targetType, t => {
+                    if (!isTupleType(t) || isTypeIdenticalTo(sourceType, t)) {
+                        return t;
+                    }
+
+                    const elementTypes: Type[] = [];
+                    const elementFlags: ElementFlags[] = [];
+
+                    for (let i = 0; i < getTypeReferenceArity(sourceType); i++) {
+                        elementTypes.push(getTupleElementType(t, i)!);

```

Comment:
It's because `getTupleElementType` guarantees a `Type` to be returned for `TupleTypeReference` arguments:
https://github.dev/microsoft/TypeScript/blob/4ba756fdc847115bc1ff5032ed06e9bed60f16b2/src/compiler/checker.ts#L22757-L22759